docs/proposals/20230612-node-os-image-caching.md

```diff
@@ -0,0 +1,225 @@
+
+In terms of data model changes, AzureMachine and AzureMachinePool changes are likely and the changes we expect will be purely additive and nonbreaking. No removals should be required to the data model. For AzureMachine on AzureMachineTemplate we will add a new optional field under spec.template.spec called nodeCachingMode which will be enabled if present and a sub field under that called cacheInterval which will be set to 1 day by default, and it will be a required subfield of nodeCachingMode if that field is there. For AzureMachinePool we will do the same thing under spec.template in the AzureMachinePool files.
+
+Example AzureMachineTemplate yaml:
+``` AzureMachineTemplate
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: AzureMachineTemplate
+metadata:
+  name: node-os-image-caching-machine-template
+  namespace: default
+spec:
+  template:
+    spec:
+      nodeCachingMode:
+        interval: 1 day
+```
+
+Example AzureMachinePool yaml:
+``` AzureMachinePool
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: AzureMachinePool
+metadata:
+  name: node-os-image-caching-machine-pool
+  namespace: default
+spec:
+  template:
+    nodeCachingMode:
+      interval: 1 day
+```
+
+A model scenario would be an operator spinning up a CAPZ cluster and having this feature be able to be toggled on or off. If it was toggled on, then as the months passed and more security updates and patches needed to be applied to the operator’s node OS image, these changes would be cached on a regular interval and the operator would no longer have to wait for these changes to apply on new node creations. This would help prevent warm nodes and overprovisioning as well as allowing the operator to have faster horizontal scaling.
+
+![Figure 1](./images/node-os-image-cache.png)
+
+### Security Model
+This proposal requires CAPZ to have write permissions for azureMachineTemplates in order to properly update the nodes’ OS image on the spec. Regex expressions will potentially be used for when a user declares a time amount for the interval that they want to specify, and denial of service attacks will be protected against by having an update system which doesn’t need to be atomic, if part of the caching is complete there is no risk in the update not finishing since the spec update will happen at once. No sensitive data is being stored in a secret. 
+
+### Risks and Mitigations
+
+Example risks:
+1. A bad snapshot is taken, and we will mitigate this risk by having a rollback if a bad snapshot is detected as taken and have prevention techniques in place to detect if the snapshot is bad before taking it.
```

Comment:
Added more details, for this proposal I'm just outlining the easiest method. A much better method would be to try to make a new node before caching for detection or some other way to prevent failure altogether, but those methods are more complex and are slower if failures are rare so the rollback is better explained instead currently since it's more in scope.